Add Unix implementation for System.Drawing.Common

mellinoe opened the pull request:
This PR adds a Unix implementation for System.Drawing.Common. This implementation has been taken from the mono codebase, and implements the majority of its image-processing and graphics functionality in a native library called "libgdiplus". This library is available on all of the Linux distributions we support, as well as macOS through homebrew. Up until the final commit, the code pulled into this repository is identical to the implementation in mono.

The final commit makes a large change to the mono codebase due to its reliance on the DllMap feature, which is not supported in .NET Core. To fix this, I have removed all of the PInvoke declarations for GDI+ functions, and replaced them with manual function loading. Please see [this commit description](https://github.com/mellinoe/corefx/commit/48e96d5ac2e51a7a388aa5d066c9f516faa22e2d) for a more thorough explanation of the process. Functionally, the behavior should be identical to before my change, but the library loading mechanism will be more flexible and resilient.

Currently, a large number of tests are disabled on Unix by this PR. A large chunk of the test failures are due to differences in argument validation or exception behavior, which can be fixed within the managed implementation. There is an unknown number of behavioral problems inherent to the libgdiplus implementation we are using, which will be much more problematic to solve. There will be a long process of understanding these behavioral differences as we go forward.

The CI machines have been updated with the new "libgdiplus" dependency, so there should be no more blockers for merging this work.

@qmfrederik @akoeplinger  @marek-safar @stephentoub 

@stephentoub I'd appreciate any input on the function-loading approach I've taken; it is in the final commit.

Fixes https://github.com/dotnet/corefx/issues/20712

marek-safar commented on src/System.Drawing.Common/src/mono/MonoTODOAttribute.cs:
```diff
@@ -0,0 +1,111 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Do you really need this file?

mellinoe replied:
There are quite a few usages in the mono code. Since we'd like to share the source files from here, I left the attributes in so that mono could still compile with them. Instead of #if'ing them out from the .NET Core version, I just defined a dummy attribute.

stephentoub replied:
Does Mono actually need them?  Some of them actually look wrong to me, like the one I commented on.

If it is imperative that we keep them, could you at least add the conditional attribute to the attribute so that it's compiled out in our release builds?

mellinoe replied:
I'd be happy to remove them if mono isn't really using or tracking them. TBH, a lot of them just say "not implemented on Unix" when the implementation already throws a NotImplementedException. So many of them aren't really tracking anything meaningful anyways...

I will definitely make them conditional for us, at the very least.

marek-safar replied:
Mono does not necessary need them. They are nice to have but as you can see they are not really updated.

stephentoub commented on src/System.Drawing.Common/src/System.Drawing.Common.csproj:
```diff
@@ -233,10 +238,202 @@
+    </Compile>
+  </ItemGroup>
+  <ItemGroup Condition="'$(TargetsUnix)' == 'true'">
+    <Compile Include="mono\System.Drawing\Bitmap.cs" />
```
Could we use a directory name other than "mono", e.g. "Unix" or something like that to match what's used elsewhere in corefx?

mellinoe replied:
Yeah, I can rename the folder.

stephentoub commented on src/System.Drawing.Common/src/System.Drawing.Common.csproj:
```diff
@@ -233,10 +238,202 @@
+    <Compile Include="mono\System.Drawing\GraphicsUnit.cs" />
+    <Compile Include="mono\System.Drawing\gdipEnums.cs" />
+    <Compile Include="mono\System.Drawing\gdipFunctions.cs" />
+    <Compile Include="System\Drawing\GdiplusNative.Unix.cs" />
```
Nit: the ordering of files here is strange

mellinoe replied:
Thank Visual Studio for that one... I will fix it.

stephentoub commented on src/System.Drawing.Common/src/System.Drawing.Common.csproj:
```diff
@@ -233,10 +238,202 @@
+    <Compile Include="mono\System.Drawing.Printing\PrintController.cs" />
+    <Compile Include="mono\System.Drawing.Printing\PrintDocument.cs">
+      <SubType>Component</SubType>
+    </Compile>
```
Nit: why do some of these have a Component subtype?

mellinoe replied:
For some reason, Visual Studio keeps adding this into the project file. I tried to remove it a couple of times. It has something to do with the base type, I think.

stephentoub commented on src/System.Drawing.Common/src/System.Drawing.Common.csproj:
```diff
@@ -233,10 +238,202 @@
+    <Compile Include="mono\System.Drawing.Printing\StandardPrintController.cs" />
+    <Compile Include="mono\System.Drawing.Text\PrivateFontCollection.cs" />
+    <Compile Include="mono\MonoTODOAttribute.cs" />
+    <!-- Shared with Windows -->
```
How about splitting the files into ItemGroups based on whether the files are Windows-specific, Unix-specific, or platform-agnostic?

mellinoe replied:
Yes, I'll do this cleanup -- then it will be more obvious what further consolidation is necessary.

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/Gdiplus.cs:
```diff
@@ -123,2380 +123,125 @@ private static void Initialize()
-            [DllImport(ExternDll.Gdiplus, SetLastError = true, ExactSpelling = true, CharSet = CharSet.Unicode)]
-            internal static extern int GdipSetPathGradientFocusScales(HandleRef brush,
-                                                                      float xScale,
-                   
```
Is Initialize called from somewhere other than the static cctor?  If not, consider just merging it into the static cctor, so that code doesn't accidentally invoke it.

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/GdiplusNative.Unix.cs:
```diff
@@ -0,0 +1,2508 @@
+            public static extern int dlclose(IntPtr handle);
+
+            [DllImport(LibName)]
+            public static extern string dlerror();
```
These P/Invokes should follow the standard conventions we've used elsewhere in the repo, e.g. these should live under src/Common/Interop/..., LibName should be in the Libraries class named appropriately, etc.

stephentoub replied:
This P/Invoke for dlerror looks wrong.  Does dlerror allocate the char* it returns and expect the caller to free it with `free`?  If it does, this is fine, but assuming not, this is going to cause problems, as the runtime will assume it owns the returned char* and try to free it as part of marshaling it to string, in which case it should instead be defined to return `char*` or `IntPtr` and Marshal.PtrToStringAnsi or Marshal.PtrToStringUTF8 should be used to get the resulting string value.

mellinoe replied:
I believe it needs to be the latter -- we shouldn't try to free the string returned from `dlerror`.

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/GdiplusNative.Unix.cs:
```diff
@@ -0,0 +1,2508 @@
+            public static extern IntPtr dlopen(string fileName, int flags);
+
+            [DllImport(LibName)]
+            public static extern IntPtr dlsym(IntPtr handle, string name);
```
Nit: we should use the same names as the platform does, e.g. name => symbol, fileName => filename, etc.

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/GdiplusNative.Unix.cs:
```diff
@@ -0,0 +1,2508 @@
+                    lib = dlopen("libgdiplus.so.0", RTLD_NOW);
+                    if (lib == IntPtr.Zero)
+                    {
+                        throw new DllNotFoundException("libgdiplus could not be loaded.");
```
This string should be in the .resx.

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/GdiplusNative.Unix.cs:
```diff
@@ -0,0 +1,2508 @@
+                IntPtr lib = dlopen("libgdiplus.so", RTLD_NOW);
+                if (lib == IntPtr.Zero)
+                {
+                    lib = dlopen("libgdiplus.so.0", RTLD_NOW);
```
I think it'd be worth a comment indicating why these two specific names were chosen, why in this order, and why not other variants.

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/GdiplusNative.Unix.cs:
```diff
@@ -0,0 +1,2508 @@
+            internal static Status GdipCombineRegionPath(IntPtr region, IntPtr path, CombineMode combineMode) => GdipCombineRegionPath_ptr.Delegate(region, path, combineMode);
+
+            private delegate Status GdipGetRegionBounds_delegate(IntPtr region, IntPtr graphics, ref RectangleF rect);
+            private static FunctionWrapper<GdipGetRegion
```
Nit: just curious, what accounts for when some lines are wrapped with indents and others not?

mellinoe replied:
Only the ones I touched manually are wrapped neatly... the rest were the raw output from a converter tool I wrote. I can see if I can format all of these automatically somehow.

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/GdiplusNative.Windows.cs:
```diff
@@ -0,0 +1,2459 @@
+            [DllImport("kernel32")]
+            private static extern IntPtr LoadLibrary(string fileName);
+            [DllImport("kernel32")]
+            private static extern IntPtr GetProcAddress(IntPtr module, string procName);
```
Same comments about being in src\Common\Interop.

stephentoub replied:
e.g. https://github.com/dotnet/corefx/blob/master/src/Common/src/Interop/Windows/kernel32/Interop.GetProcAddress.cs

mellinoe replied:
Thanks, I will relocate these (or re-use the existing definition if it's already there).

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/GdiplusNative.Windows.cs:
```diff
@@ -0,0 +1,2459 @@
+            [DllImport("kernel32")]
+            private static extern IntPtr GetProcAddress(IntPtr module, string procName);
+
+            private static IntPtr LoadNativeLibrary() => LoadLibrary("gdiplus.dll");
```
Why does the Windows implementation not just use the same P/Invokes it always has?

mellinoe replied:
See longer comment down below

stephentoub commented on src/System.Drawing.Common/src/mono/System.Drawing.Drawing2D/GraphicsContainer.cs:
```diff
@@ -0,0 +1,54 @@
+    public sealed class GraphicsContainer : MarshalByRefObject
+    {
+
+        private uint nativeState;
```
Nit: can we make the style in these imported files better match the rest of corefx? e.g. `private readonly uint _nativeState`, or even using `internal uint NativeObject { get; }`.

stephentoub commented on src/System.Drawing.Common/src/mono/System.Drawing.Drawing2D/GraphicsPath.cs:
```diff
@@ -0,0 +1,896 @@
+
+		GraphicsPath (IntPtr ptr)
+		{
+			nativePath = ptr;
```
It appears tabs are being used instead of spaces.  Can you run the codeformatter?

mellinoe replied:
Hmm, yes. I had run the formatter (or at least Visual Studio document formatting) on most files, but it appears I've missed some.

stephentoub commented on src/System.Drawing.Common/src/mono/System.Drawing.Drawing2D/GraphicsPath.cs:
```diff
@@ -0,0 +1,896 @@
+		public GraphicsPath (Point[] pts, byte[] types, FillMode fillMode)
+		{
+			if (pts == null)
+				throw new ArgumentNullException ("pts");
```
nameof

stephentoub commented on src/System.Drawing.Common/src/mono/System.Drawing.Drawing2D/GraphicsPath.cs:
```diff
@@ -0,0 +1,896 @@
+		public GraphicsPath ()
+		{
+                        Status status = SafeNativeMethods.Gdip.GdipCreatePath (FillMode.Alternate, out nativePath);
+                        SafeNativeMethods.Gdip.CheckStatus (status);
```
I'm unclear on something: if libgdiplus is meant to provide the same surface area and behavior as gdiplus, why does anything about the interop layer differ between platforms? i.e. why does this file exist at all?

mellinoe replied:
See longer comment below

stephentoub commented on src/System.Drawing.Common/src/mono/System.Drawing.Drawing2D/GraphicsPath.cs:
```diff
@@ -0,0 +1,896 @@
+			if (pts == null)
+				throw new ArgumentNullException ("pts");
+			if (pts.Length != types.Length)
+				throw new ArgumentException ("Invalid parameter passed. Number of points and types must be same.");
```
Resource strings should be in the resx.  Will stop commenting on this.

stephentoub commented on src/System.Drawing.Common/src/mono/System.Drawing.Drawing2D/GraphicsPath.cs:
```diff
@@ -0,0 +1,896 @@
+		public GraphicsPath (PointF[] pts, byte[] types, FillMode fillMode)
+		{
+			if (pts == null)
+				throw new ArgumentNullException ("pts");
```
Nit: nameof.  Will stop commenting on this.

stephentoub commented on src/System.Drawing.Common/src/mono/System.Drawing.Drawing2D/GraphicsPath.cs:
```diff
@@ -0,0 +1,896 @@
+                        Dispose (false);
+                }
+                
+		void Dispose (bool disposing)
```
private

stephentoub commented on src/System.Drawing.Common/src/mono/System.Drawing.Drawing2D/GraphicsPath.cs:
```diff
@@ -0,0 +1,896 @@
+		}
+
+		public FillMode FillMode {
+			get {
```
Nit: brace style doesn't match that used elsewhere in corefx.  Hopefully codeformatter would address that.

stephentoub commented on src/System.Drawing.Common/src/mono/System.Drawing.Drawing2D/GraphicsPath.cs:
```diff
@@ -0,0 +1,896 @@
+			}
+			set {
+				nativePath = value;
+			}
```
Nit: `internal IntPtr NativeObject { get; set; }`?

stephentoub commented on src/System.Drawing.Common/src/mono/System.Drawing.Drawing2D/GraphicsPath.cs:
```diff
@@ -0,0 +1,896 @@
+                        SafeNativeMethods.Gdip.CheckStatus (status);                      	
+                }
+                
+		[MonoTODO ("The StringFormat parameter is ignored when using libgdiplus.")]
```
Do we need these attributes?

stephentoub replied:
Also, this doesn't appear to be ignored by the C# code; if anything is ignoring it, it would seem to be libgdiplus itself.

stephentoub commented:
It looks like the change adds a bunch of code and then removes a bunch of that code in subsequent commits.  It'd be nice to squash this when merging so that such temporary code doesn't pollute the repo's history.

mellinoe commented:
>It looks like the change adds a bunch of code and then removes a bunch of that code in subsequent commits. It'd be nice to squash this when merging so that such temporary code doesn't pollute the repo's history.

Yeah, I'll do that. The branch looks like it does currently because some consolidation commits were done over in the mono repo after I pulled in the initial snapshot. And I just pulled those commits over individually.

mellinoe commented:
@stephentoub You made a few comments regarding the split between Windows/Linux PInvokes, why we aren't just using the old PInvokes on Windows, etc.

Ultimately, I would like to get the codebase into a place where all of the native imports are shared, and there is only a single declaration for each function. You'll see that some of the declarations are shared right now, and some aren't. The reason for that is @qmfrederik already did some preliminary consolidation of the codebases over in the mono repo. When he did that, he switched the mono version to use the Windows codebase, and therefore the Windows PInvoke declarations. We should be able to continue doing that consolidation and deleting the duplicated function imports.

Although libgdiplus is signature-compatible, it isn't 100% behavior compatible. Some function do seem to "blow up" if you don't validate parameters up-front (which the mono version seems to do in some cases). Also, there are some "extra" functions (suffixed with "_linux", "_macos", etc.) that provide platform-specific hooks for some parts of the code. I think that even in the end we will probably keep chunks of the mono code, probably in partial class files, that handle those little differences. But the majority of logic, as well as all of the function imports, should be shared.

You do have a good general point, though: ultimately we will (hopefully) delete a good chunk of the mono code, when we consolidate (and test) it with the Windows implementation. I felt that it was an easier workflow to

1) Get things merged in a working state with mono's code, even with some code duplication
2) Consolidate the two implementations with the help from the test cases we have checked in now

Does that seem like a reasonable approach?

stephentoub commented:
> Does that seem like a reasonable approach?

Yup, thanks.

mellinoe commented:
* Rebased the branch to consolidate the initial add-and-delete commits
* Rebased out some of the intermediate stuff so that the commits are more clear
* Did some additional "basic" cleanup of the mono code, including running the VS formatter. Mainly this just cleaned up some remaining tabs in the source code, and incorrect brace style.
* Cleaned up the project file significantly
* Moved around and consolidated the interop code so that it matches the guidelines more closely.
* Moved the  "Initialize" method into the static constructor (it was only ever allowed to be called from the cctor anyways).
* Added a "LibgdiplusNotFound" string resource used by the Unix function loading code.
* Added a `ConditionalAttribute` to the MonoTODOAttribute class, such that it won't be present in our built assemblies. We can probably just delete these most likely, but I will leave that for later.

I'm going to merge this as-is now, so we can start tracking the test behavior against this implementation and start really consolidating the implementations. There's some more obvious cleanup that should be done as we consolidate the code:

* Switch to more corefx-friendly code style, including using `nameof`, embedded string resources, etc. In a lot of cases, we should be able to share the argument validation and exception throwing logic with Windows, so many of these problems will go away from the consolidation work.
* Move source files into the "regular source tree" as they are consolidated, rather than in the "Unix" subdirectory.
